### ISSUE TITLE
unable to connect to cluster from Lens Terminal, if WSL is selected.

### ISSUE THREAD

**Issue by jkroepke** [kontena/lens#6021]
**Describe the bug**
Open the Lens terminal and running kubectl reports an error:

```
The connection to the server 127.0.0.1:58109 was refused - did you specify the right host or port?
```

And i'm not able to connect to the cluster from lens terminal. Pod Shells doesnt work, too.

**To Reproduce**
1. Start Lens on a Windows system with WSL enabled
2. Click preferences Terminal, enter 'C:\Windows\System32\wsl.exe'
3. Connect to a cluster and open Terminal

**Expected behavior**
kubectl and pod shells should work

**Screenshots**
N/A

**Environment (please complete the following information):**
- Lens Version: 6.0.0
- OS: [e.g. OSX] Windows
- Installation method (e.g. snap or AppImage in Linux): exe

**Logs:**
When you run the application executable from command line you will see some logging output. Please paste them here:
```
Your logs go here...
```

**Kubeconfig:**
Quite often the problems are caused by malformed kubeconfig which the application tries to load. Please share your kubeconfig, remember to remove any secret and sensitive information.
```
your kubeconfig here
```

**Additional context**
The issue here is that Lens is using an auth proxy to talk with the cluster. The auth proxy runs on 127.0.0.1. Since WSL2 is a dedicated virtual machine, 127.0.0.1 is different from WSL perspective. 

See also: https://github.com/microsoft/WSL/issues/4619

2 idea to solve this:

1. make auth proxy listen ip configurable. This is terrible if the auth proxy is reachable through network.
2. Generate a kubeconfig without auth-proxy, but make sure the correct kubeconfig with the correct context is set.

**Comment by Hackmeat** [comment]
is their a fast fix for it? ran into the exact same issue

**Comment by kobethuwis** [comment]
Any progress/fixes? Have the same issue.

**Comment by jkroepke** [comment]
[WSLg](https://github.com/microsoft/wslg) is maybe a workaround here, since Lens will completely run inside WSL.

**Comment by RonNewcomb** [comment]
same issue.

**Comment by samueljholiver** [comment]
same issue here!

**Comment by kylebisley** [comment]
I was able to solve this by downgrading WSL.

In poweshell run
`wsl -l -v`
wait for it to complete. Ensure that you've got the wsl.exe terminal selected in lens. Lens->File->Preferences->Terminal to ensure Terminal Shell Path reads `wsl.exe`

I believe it's linked to a wider WSL2 issue around localhost differences between WSL and Windows though an interaction with len's proxy. I was unable to get other workarounds to function though so downgrading is my route for now.

Hope this helps someone.

**Comment by hyprsnpr** [comment]
I mostly found a solution for this. Basically 127.0.0.1 is being added by lens, WSL can't access a hosts localhost. However the kubeconfig's that get added to lens have the correct url's in them. So if you keep all your kubeconfig's in a common directory structure you can get a little clever using aliases in the .bashrc.

Add these lines to your ~/.bashrc:
alias kubectl='kubectl --kubeconfig=$(grep -rl --include='*.yml' --include='*.yaml' "$(cat $KUBECONFIG | grep name: | head -n 1 | awk '\''{ print $3 }'\'')" /mnt/c/path/to/your/kubeconfigs/)'

alias helm='helm --kubeconfig=$(grep -rl --include='*.yml' --include='*.yaml' "$(cat $KUBECONFIG | grep name: | head -n 1 | awk '\''{ print $3 }'\'')" /mnt/c/path/to/your/kubeconfigs/)'

Then reload your .bashrc
source ~/.bashrc

Hope this helps.

**Comment by igorbarannic** [comment]
@hyprsnpr I've found another workaround if you have one file with kube configs
1. set in Lens > Preferences > Terminal > path to wsl (e.g. wsl.exe)
2. Inside your wsl edit ~/.bashrc and add nextline
  alias kubectl="KUBECONFIG=/root/.kube/config; kubectl" # /root/.kube/config - path to config file
3. After re-opening Lens you will have access to your cluster in Lens terminal window

**Comment by nemke43** [comment]
@igorbarannic I also had to disable following:
1. Lens > Preferences > Kuberentes > Download kubectl binaries matching the Kuberenets cluster version 
2. leave PATH TO KUBECTL BINAY field empty

Rest is same as yours:
 
3. set in Lens > Preferences > Terminal > path to wsl (e.g. wsl.exe)
4. Inside your wsl edit ~/.bashrc and add nextline
    alias kubectl="KUBECONFIG=/root/.kube/config; kubectl" # /root/.kube/config - path to config file
5. After re-opening Lens you will have access to your cluster in Lens terminal window

**Comment by sanchpet** [comment]
Unfortunately, these workarounds do not help if you manage more than 1 cluster, because the context is not switching automatically.

**Comment by sanchpet** [comment]
This is my solution that can help if you manage multiple clusters in same kubeconfig.

Add in `.bashrc`/`.zshrc` following:
```
# If set, means we run console from Lens
if [[ -n "$KUBECONFIG" ]]; then
        # Save lens kubeconfig to parse context
        LENS_KC=$KUBECONFIG
        # Reset kubeconfig to manipulate our actual kubeconfig with kubectl
        KUBECONFIG=
        # Use context that wat put in Lens kubeconfig
        kubectl config use-context $(cat $LENS_KC | yq '.contexts.[].name' | sed 's/"//g') > /dev/null
fi
```

Initially `$KUBECONFIG` is empty, but if we run WSL from Lens, it's prefilled. So if it's the case, we save "lens kubeconfig" to a new variable, then reset `$KUBECONFIG` to manipulate actual config with kubectl. After that we parse context name and use it in our config from WSL. 

Basically on terminal start, if Lens sets kubeconfig, we drop it and set same context that's opened in Lens.

This is useful, for example, for `yc` (Yandex Cloud) tool where you have to add multiple clusters into your kubeconfig.
